File: C++/lintcode_cpp/464_sort_integers_ii.cpp
// // solution1, quicksort
// psedo code:
// quickSort(arr[], low, high)
// {
//     if (low < high)
//     {
//         /* pi is partitioning index, arr[pi] is now
//           at right place */
//         pi = partition(arr, low, high);

//         quickSort(arr, low, pi - 1);  // Before pi
//         quickSort(arr, pi + 1, high); // After pi
//     }
// }
// class Solution {
// public:
//     /**
//      * @param A: an integer array
//      * @return: nothing
//      */
//     void sortIntegers2(vector<int> &A) {
//         // write your code here
//         quickSort(A, 0, A.size() - 1);
//     }
    
//     void quickSort(vector<int> &A, int left, int right) {
//         if (left < right) {
//             int idx = partition(A, left, right, (left+right)/2);
//             quickSort(A, left, idx - 1);
//             quickSort(A, idx + 1, right);
//         }
//     }
    
//     // do not miss out the left, right in partition!!!!!!!!!!1
//     // set idx = left instead of 0!!!!
//     int partition(vector<int> &A, int left, int right, int pivot) {
//         int val = A[pivot];
//         swap(A, pivot, right);
//         int idx = left;
//         for (int i = left; i <= right; i++) {
//             if (A[i] < val) {
//                 swap(A, idx, i);
//                 idx++;
//             }
//         }
//         swap(A, idx, right);
//         return idx;
//     }
    
//     void swap(vector<int> &A, int a, int b) {
//         int temp = A[a];
//         A[a] = A[b];
//         A[b] = temp;
//     }
// };


// Solution 2, merge sort
// MergeSort(arr[], l,  r)
// If r > l
//      1. Find the middle point to divide the array into two halves:  
//              middle m = (l+r)/2
//      2. Call mergeSort for first half:   
//              Call mergeSort(arr, l, m)
//      3. Call mergeSort for second half:
//              Call mergeSort(arr, m+1, r)
//      4. Merge the two halves sorted in step 2 and 3:
//              Call merge(arr, l, m, r)
class Solution {
public:
    /**
     * @param A: an integer array
     * @return: nothing
     */
    void sortIntegers2(vector<int> &A) {
        mergeSort(A, 0, A.size() - 1);
    }
    
    void mergeSort(vector<int> &A, int left, int right) {
        // base case: 0 or 1 element in array, return
        if (left < right) {
            int mid = (left + right) / 2;
            // divide
            mergeSort(A, left, mid);
            mergeSort(A, mid + 1, right);
            merge(A, left, right, mid);
        }
    }
    
    void merge(vector<int> &A, int left, int right, int mid) {
        int i = left, j = mid + 1, k = 0, temp[right-left+1];
        while (i <= mid && j <= right) {
            if (A[i] < A[j]) {
                temp[k] = A[i];
                i++;
            } else {
                temp[k] = A[j];
                j++;
            }
            k++;
        }
        while (i <= mid) {
            temp[k++] = A[i++];
        }
        while (j <= right) {
            temp[k++] = A[j++];
        }
        for (int i = left; i <= right; i++) {
            A[i] = temp[i-left];
        }
    }
};

File: Python/lintcode_python/464_sort_integers_ii.py

# # merge sort

# class Solution:
#     """
#     @param A: an integer array
#     @return: nothing
#     """

#     def sortIntegers2(self, A):
#         self.mergeSort(A, 0, len(A) - 1)
#         return A

#     def mergeSort(self, A, left, right):
#         if left >= right:
#             return A
#         if left == right - 1:
#             if A[left] > A[right]:
#                 A[left], A[right] = A[right], A[left]
#             return
#         partition = (left + right) // 2
#         self.mergeSort(A, left, partition)
#         self.mergeSort(A, partition + 1, right)
#         self.merge(A, left, right, partition)
#         return

#     def merge(self, A, left, right, partition):
#         if left >= right:
#             return
#         temp = [0 for _ in range(right-left+1)]
#         A_left = A[left:partition + 1]
#         A_right = A[partition + 1:right+1]

#         idx = l_idx = r_idx = 0
#         while l_idx < len(A_left) and r_idx < len(A_right):
#             if A_left[l_idx] < A_right[r_idx]:
#                 temp[idx] = A_left[l_idx]
#                 l_idx += 1
#             else:
#                 temp[idx] = A_right[r_idx]
#                 r_idx += 1
#             idx += 1

#         if l_idx < len(A_left):
#             # temp.extend(A_left[l_idx:len(A_left)]) # bug
#             temp[idx:] = A_left[l_idx:len(A_left)]
#         else:
#             temp[idx:] = A_right[r_idx:len(A_right)]

#         A[left:right + 1] = temp
#         return




# Solution 1, merge sort
# class Solution:
#     """
#     @param A: an integer array
#     @return: nothing
#     """
#     def sortIntegers2(self, A):
#         if not A:
#             return
#         self.mergeSort(A, 0, len(A) - 1)
        
#     def mergeSort(self, A, left, right):
#         if left < right:
#             mid = (left + right) // 2
#             self.mergeSort(A, left, mid)
#             self.mergeSort(A, mid + 1, right)
#             self.merge(A, left, right, mid)
        
#     def merge(self, A, left, right, mid):
#         if left < right:
#             i, j = left, mid + 1
#             k = 0
#             temp = [0 for _ in range(right - left + 1)]
            
#             while i <= mid and j <= right:
#                 if A[i] < A[j]:
#                     temp[k] = A[i]
#                     i += 1
#                 else:
#                     temp[k] = A[j]
#                     j += 1
#                 k += 1
            
#             while i <= mid:
#                 temp[k] = A[i]
#                 i += 1
#                 k += 1
            
#             while j <= right:
#                 temp[k] = A[j]
#                 j += 1
#                 k += 1
#             for i in range(right - left + 1):
#                 A[i+left] = temp[i]
                
            
        
        
# Solution 2, quick sort
class Solution:
    """
    @param A: an integer array
    @return: nothing
    """
    def sortIntegers2(self, A):
        self.quickSort(A, 0, len(A) - 1)
        
    def quickSort(self, A, left, right):
        if right > left: # otherwise maximum recursion exceed
            p = self.partition(A, left, right, (left + right) // 2)
            self.quickSort(A, left, p - 1)
            self.quickSort(A, p + 1, right)

    def partition(self, A, left, right, pivot):
        p_val = A[pivot]
        A[right], A[pivot] = A[pivot], A[right]
        
        idx = left
        # for i in range(left, right - 1) @ wrong
        for i in range(left, right):
            if A[i] < p_val:
                A[idx], A[i] = A[i], A[idx]
                idx += 1
        A[idx], A[right] = A[right], A[idx]
        return idx
